token-tracker/src/truth-sources/lp-tokens.js

```diff
@@ -0,0 +1,273 @@
+   *
+   * @param {Array<Address>} lpStakers LP KEEP-ETH / KEEP-TBTC stakers
+   *
+   * @return {Map<Address,BN>} LP Balances by lp stakers
+   */
+  async getLpTokenStakersBalances(lpStakers) {
+    const lpBalanceByStaker = new Map()
+    let expectedTotalSupply = new BN(0)
+
+    for (let i = 0; i < lpStakers.length; i++) {
+      const lpBalance = new BN(
+        await this.liquidityStaking.lpRewardsContract.methods
+          .balanceOf(lpStakers[i])
+          .call({}, this.targetBlock)
+      )
+      if (!lpBalance.isZero()) {
+        lpBalanceByStaker.set(lpStakers[i], lpBalance)
+        expectedTotalSupply = expectedTotalSupply.add(lpBalance)
+      }
+    }
+    const actualTotalSupply = new BN(
+      await this.liquidityStaking.lpRewardsContract.methods
+        .totalSupply()
+        .call({}, this.targetBlock)
+    )
+
+    if (!expectedTotalSupply.eq(actualTotalSupply)) {
+      logger.error(
+        `Sum of LP staker balances ${expectedTotalSupply} does not match the total supply ${actualTotalSupply}`
+      )
+    }
+
+    logger.info(`Total supply of LP Token: ${expectedTotalSupply.toString()}`)
+
+    return lpBalanceByStaker
+  }
+
+  /**
+   * Calculates KEEP for all LP KEEP-ETH / KEEP-TBTC stakers.
+   *
+   * @param {Map<Address, BN>} stakersBalances LP KEEP-ETH / KEEP-TBTC Token amounts by stakers
+   *
+   * @return {Map<Address,BN>} KEEP Tokens in LP KEEP-ETH / KEEP-TBTC at the target block
+   */
+  async calcKeepInStakersBalances(stakersBalances) {
+    logger.info(`check token stakers at block ${this.targetBlock}`)
+
+    const keepInLpByStakers = new Map()
+
+    // Retrieve current pair data
+    const pairData = await getPairData(this.liquidityStaking.lpPairAddress)
+    for (const [stakerAddress, lpBalance] of stakersBalances.entries()) {
+      const keepInLPToken = await this.calcKeepTokenfromLPToken(
+        lpBalance,
+        pairData
+      )
+      keepInLpByStakers.set(stakerAddress, keepInLPToken)
+
+      logger.info(
+        `Staker: ${stakerAddress} - LP Balance: ${lpBalance} - KEEP in LP: ${keepInLPToken}`
+      )
+    }
+
+    logger.info(
+      `found ${keepInLpByStakers.size} stakers at block ${this.targetBlock}`
+    )
+
+    dumpDataToFile(
+      keepInLpByStakers,
+      this.liquidityStaking.keepInLpTokenFilePath
+    )
+
+    return keepInLpByStakers
+  }
+
+  /**
+   * Calculates amount of KEEP token which makes a KEEP-ETH / KEEP-TBTC Uniswap pair.
+   * A Uniswap LP pair is a bookeeping tool to keep track of how much the liquidity
+   * stakers are owed. They store two assets of equivalent value of each, ex. KEEP-ETH.
+   * This means that the value of KEEP owned is dependent on the ratio of staked
+   * LP tokens and the total LP supply. Ratio between LP tokens and KEEP tokens
+   * should be equal:
+   * LP_staker_balance / LP_total_supply_pool == KEEP_staker_owed / KEEP_total_liquidity_pool
```

Comment:
Just a note: it can also be explained based on [how LP tokens are minted](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2Pair.sol#L123):

```
lp = token_amount_deposited * total_LP_supply / token_amount_in_reserve
```
Thus:
```
token_amount_deposited = lp * token_amount_in_reserve / total_LP_supply
```

Reply:
Added a reference to description: 5ad059d